### PR TITLE
Lovelace Layout: Drag and Drop

### PR DESCRIPTION
**Pull request by zsarnett**
## Proposed change
Adding the ability to drag and drop cards to an order. The order is determined by a left to the right path. Cards are placed into columns by the order they come in Lovelace configuration

[Screenshot from Gyazo](https://gyazo.com/825392c07ac6501a74d39bd0aab8d8ba)

## Type of change

- Dependency upgrade
- Bugfix (non-breaking change which fixes an issue)
- [x] New feature (thank you!)
- [x] Breaking change (fix/feature causing existing functionality to break)
- Code quality improvements to existing code or addition of tests

## Additional information

To-Do:
- [x] Remove comments
- [x] Only allow dragging when in edit mode
- [ ] Update the method of saving the Lovelace Config (Assistance Needed)
- [x] Find a solution to refreshing the items when the height is determined for each card (Assistance needed)
- [x] Type out all objects?
- [x] Is this the best way to import the Murri Library?
- [x] Update the shadow when the item is being moved

- This PR is related to issue: https://github.com/home-assistant/home-assistant-polymer/issues/3895

## Checklist

- [x] The code change is tested and works locally.
- [x] There is no commented out code in this PR.

### CODE REVIEW

**Comment by zsarnett** [comment]
I found that someone did some of the typings that we may need

https://gist.github.com/sploders101/84f88e0711ad573c3cc538ce112a98c8

**Comment by zsarnett** [comment]
As you can see in the gif, When the first render processes the Murri grid doesn't know the height of the cards... This is why you will see overlap, and then me clicking a badge which triggers the `updated` method that runs `grid.refreshItems().layout()` on each grid. Need to find a way to not show the cards until the height is found or immediately do the refresh without running the updated method

**Comment by zsarnett** [comment]
Biggest issue right now seems to be the History Graph card because it will load after the updated waiting on the data to load in the card

**Comment by zsarnett** [comment]
Also, experimenting with the idea of not only making them not draggable during non-edit mode but not creating the grids at all. Storing the column elements in an array and appending the cards to each column instead of adding to the grid.

**Comment by zsarnett** [comment]
I know it isn't Ideal but I was able to fix the resizing issues when using Resize Observer to observe the changes in height and to refresh the items when this happens (Not pushed). Not sure if there is a better method. Maybe when the cards finish loading their data they can update the view that the finished?

File: package.json [hunk 1]
```diff
@@ -96,6 +96,7 @@
     "mdn-polyfills": "^5.16.0",
     "memoize-one": "^5.0.2",
     "moment": "^2.24.0",
+    "muuri": "^0.8.0",
```
Comment (bramkragten) [inline package.json]: Missing `yarn.lock`

File: src/panels/lovelace/views/hui-view.ts [hunk 2]
```diff
@@ -5,6 +5,7 @@ import {
   TemplateResult,
   property,
 } from "lit-element";
+import Muuri from "muuri/dist/muuri";
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: We should only import it in edit mode

File: src/panels/lovelace/views/hui-view.ts [hunk 3]
```diff
@@ -176,11 +160,45 @@ export class HUIView extends LitElement {
...
+        .grid {
+          position: relative;
+          width: 100%;
+        }
+
+        .item {
+          display: block;
+          width: 100%;
+          margin: 5px;
+          z-index: 1;
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: Move this to `.item.edit` and maybe something like this:
```
          box-sizing: border-box;
          padding: 4px;
          width: 100%;
          z-index: 1;
```

File: src/panels/lovelace/views/hui-view.ts [hunk 4]
```diff
@@ -176,11 +160,45 @@ export class HUIView extends LitElement {
...
+          position: relative;
+          width: 100%;
+          height: 100%;
+        }
       </style>
     `;
   }
 
-  protected updated(changedProperties: PropertyValues): void {
+  protected async updated(changedProperties: PropertyValues): Promise<void> {
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: updated can't be `async`, make it `this.updateComplete.then(() => { /* do stuff */ });`

File: package.json (hunk 1 above)
Comment (zsarnett) [reply package.json]: ah I thought this was something that wasn't push..

File: src/panels/lovelace/views/hui-view.ts [hunk 5]
```diff
@@ -284,49 +308,98 @@ export class HUIView extends LitElement {
...
+        dragReleaseEasing: "ease",
+        dragSortInterval: 0,
+        layoutDuration: 400,
+        layoutEasing: "ease",
+      };
+
+      columns.forEach((columnEl) => {
+        // Push each grid per column
+        this._grids.push(
+          new Muuri(columnEl, options) // The events here make sure the card doesnt grow to the size of the window
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: Can we only use Muuri in `editMode`?

File: src/panels/lovelace/views/hui-view.ts (hunk 4 above)
Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: Gotcha. I figured this wasn't ideal. Thank you for the alt method

File: src/panels/lovelace/views/hui-view.ts [hunk 6]
```diff
@@ -351,6 +424,50 @@ export class HUIView extends LitElement {
       curBadgeEl === badgeElToReplace ? newBadgeEl : curBadgeEl
     );
   }
+
+  private _storeLayout(): void {
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: Instead of going over all the card on every drag, can't we just move the card we just dropped?

Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: That would be ideal. Just need to find the correct algorithm that finds where the card should be placed and "slices" the card into the config. Something I haven't put much time in yet. Something I need some help with is what to do when saving the lovelace configuration. Saving it causes an update that makes the cards refresh and replace them (same order but cards go from zero height to full height. When I remove the save of the config this issue doesn't exist. Potentially a save placement button?

Comment (iantrich) [reply src/panels/lovelace/views/hui-view.ts]: > Potentially a save placement button?
Yeah, just like our card configs, not reason to save anything until after the user is done messing about

File: src/panels/lovelace/views/hui-view.ts (hunk 2 above)
Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: What is the best way to do this and where? Dynamic imports are async. so not sure if we put the import in `updated` and then update ifs to `if imported and in edit mode`

**Comment by zsarnett** [comment]
Last Commit adds Resize Observer. Let me know your thoughts on if this is the best route or if you know of some point in the element lifecycle we can listen for when the element is resized (ex. history graph - its async and resizes after the card is placed)

File: src/panels/lovelace/views/hui-view.ts (hunk 6 above)
Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: If this is the route we want to go where would the save button go? If a card is moved replace the + action fab with a Save icon?

Add a second fab above the add card fab that appears if a card is moved? (Material Design says this is okay if the actions are both important. Three is bad)

Add a button to the top of the app header? This would take adding configuration that is passed back to the header that a card was moved.

Whatever we choose I think the best option would not to show this until a card has been moved

**Comment by zsarnett** [comment]
Should I add the original code back and make this layout method optional? or should the new layout method be for everyone

File: src/panels/lovelace/views/hui-view.ts (hunk 2 above)
Comment (bramkragten) [reply src/panels/lovelace/views/hui-view.ts]: Add it to <https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/panels/lovelace/views/hui-view-editable.ts>

File: src/panels/lovelace/views/hui-view.ts (hunk 6 above)
Comment (bramkragten) [reply src/panels/lovelace/views/hui-view.ts]: We should not have a save button, it makes the user experience bad, you should just prevent the update from happening here <https://github.com/home-assistant/home-assistant-polymer/blob/dev/src/panels/lovelace/views/hui-view.ts#L215>.
Btw, what happens when a card changes size due to a state update (conditional) while in edit mode?

Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: @bramkragten Checkout my comments about that on this PR 😉

File: src/panels/lovelace/views/hui-view.ts [hunk 7]
```diff
@@ -5,6 +5,8 @@ import {
   TemplateResult,
   property,
 } from "lit-element";
+import Muuri from "muuri/dist/muuri";
+import { ResizeObserver } from "resize-observer";
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: Use it as a Polyfill, so use the native if it is available.

**Comment by zsarnett** [comment]
Not sure if any on this is correct for the Polyfill or import. Would love feedback on the best practice here.

Polyfill - I used code from the map card. I removed the polyfill of the package and used the event listener

Import - Not sure the best way to wait on the import to finish. I made the code loaded a property to listen for changes


Also, I would like some help with not recreating the cards when Lovelace is saved. It seems that the view is recreated when Lovelace is saved therefore all properties are reset to undefined

File: src/panels/lovelace/views/hui-view-editable.ts [hunk 8]
```diff
@@ -1,3 +1,5 @@
 // hui-view dependencies for when in edit mode.
 import "../components/hui-card-options";
 import "../../../components/ha-fab";
+import Muuri from "muuri/dist/muuri";
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view-editable.ts]:
```suggestion
export Muuri from "muuri/dist/muuri";
```

File: src/panels/lovelace/views/hui-view.ts [hunk 9]
```diff
@@ -219,9 +232,13 @@ export class HUIView extends LitElement {
...
     const lovelace = this.lovelace!;
 
-    if (lovelace.editMode && !editCodeLoaded) {
-      editCodeLoaded = true;
-      import(/* webpackChunkName: "hui-view-editable" */ "./hui-view-editable");
+    if (lovelace.editMode && !this.editCodeLoaded) {
+      import(
+        /* webpackChunkName: "hui-view-editable" */ "./hui-view-editable"
+      ).then((muuri) => {
+        murriGrid = muuri.default;
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]:
```suggestion
        murriGrid = editCode.Muuri;
```

File: src/panels/lovelace/views/hui-view.ts [hunk 10]
```diff
@@ -219,9 +232,13 @@ export class HUIView extends LitElement {
     const hass = this.hass!;
     const lovelace = this.lovelace!;
 
-    if (lovelace.editMode && !editCodeLoaded) {
-      editCodeLoaded = true;
-      import(/* webpackChunkName: "hui-view-editable" */ "./hui-view-editable");
+    if (lovelace.editMode && !this.editCodeLoaded) {
+      import(
+        /* webpackChunkName: "hui-view-editable" */ "./hui-view-editable"
+      ).then((muuri) => {
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]:
```suggestion
      ).then((editCode) => {
```

File: src/panels/lovelace/views/hui-view.ts [hunk 11]
```diff
@@ -219,9 +232,13 @@ export class HUIView extends LitElement {
     const hass = this.hass!;
     const lovelace = this.lovelace!;
 
-    if (lovelace.editMode && !editCodeLoaded) {
-      editCodeLoaded = true;
-      import(/* webpackChunkName: "hui-view-editable" */ "./hui-view-editable");
+    if (lovelace.editMode && !this.editCodeLoaded) {
```
Comment (bramkragten) [inline src/panels/lovelace/views/hui-view.ts]: Don't move `editCodeLoaded` to the class

**Comment by bramkragten** [comment]
> Polyfill - I used code from the map card. I removed the polyfill of the package and used the event listener

The is no polyfill now, you can do this (also import `resize-observer` only when in edit mode):
```ts
import { install } from 'resize-observer';
 
if (!window.ResizeObserver) install();
```

File: src/panels/lovelace/views/hui-view.ts (hunk 11 above)
Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: I am not sure how to let the View know that the code was loaded and to only try and load edit mode once that is done without this

File: src/panels/lovelace/views/hui-view-editable.ts (hunk 8 above)
Comment (zsarnett) [reply src/panels/lovelace/views/hui-view-editable.ts]: I don't think this is valid, is it? I receive many errors not just for the types not being included

File: src/panels/lovelace/views/hui-view.ts [hunk 12]
```diff
@@ -212,7 +263,24 @@ export class HUIView extends LitElement {
...
+        return;
+      }
+      this._resizeObserver.disconnect();
+    }
+
+    if (
+      configChanged ||
+      editModeChanged ||
+      changedProperties.has("columns") ||
+      changedProperties.has("editCodeLoaded")
```
Comment (zsarnett) [inline src/panels/lovelace/views/hui-view.ts]: TODO: Figure out the logic to not recreate the cards in edit mode when cards are moved

File: src/panels/lovelace/views/hui-view.ts [hunk 13]
```diff
@@ -20,31 +20,28 @@ import { HomeAssistant } from "../../../types";
...
-    }
-    if (columnEntityCount[i] < columnEntityCount[minIndex]) {
-      minIndex = i;
-    }
-  }
-
-  columnEntityCount[minIndex] += size;
-
-  return minIndex;
+let muuri: any;
```
Comment (zsarnett) [inline src/panels/lovelace/views/hui-view.ts]: TODO: Import types from d.ts files. I think I may need to update TSConfig to look at types in a local folder?

File: src/panels/lovelace/views/hui-view.ts [hunk 14]
```diff
@@ -351,6 +455,63 @@ export class HUIView extends LitElement {
...
+        if (!grid.getItems(i)[0]) {
+          return;
+        }
+
+        layout.push(grid.getItems(i)[0].getElement().cardConfig);
+      });
+    }
+
+    // Save Lovelace config - Yes I know - This isnt great
+    this.lovelace!.config.views.forEach((viewConf, index) => {
```
Comment (zsarnett) [inline src/panels/lovelace/views/hui-view.ts]: Only save the card that was moved?

**Comment by zsarnett** [comment]
> (Make Columns a grid after making the cards  1b9b621)

- Updated Gif in Original Post
- Instead of creating the grids and then appending the cards into the grids. Keep the original way for adding all cards to columns and then after the cards have taken shape (height changed) then turn the columns into Grids. This gets rid of the initial layout issues where cards would show 8px in height and then expand.

File: src/panels/lovelace/views/hui-view.ts (hunk 14 above)
Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: I am not sure that it will be possible to only save the card that was moved. Because the cards are laid out 1 card per column then wrap to the first column but when the drag and drop happens the cards are in columns and shift up and down. Changing the position of one card will change many cards position.

File: src/panels/lovelace/views/hui-view.ts (hunk 13 above)
Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts]: I have imported the Types from the Link in the PR above. I have added them to a relative @types path. Not sure this is where we want this? 

No need to update the TSConfig File. Its so smart :) 

How can I type the Muuri variable in the view file without importing Muuri

**Comment by zsarnett** [comment]
@bramkragten  This is something I need help with. I think that this is basically done. Except for how views are recreated every time the Lovelace config is changed. I need the view to stay intact in order to persist the property values

File: src/panels/lovelace/types.ts [hunk 15]
```diff
@@ -19,9 +19,10 @@ export interface Lovelace {
   editMode: boolean;
   mode: "generated" | "yaml" | "storage";
   language: string;
+  rebuild: string;
```
Comment (zsarnett) [inline src/panels/lovelace/types.ts]: Should this be a boolean?

Comment (bramkragten) [reply src/panels/lovelace/types.ts]: Yes

**Comment by zsarnett** [comment]
@bramkragten Here is where the config changes and `force` is set to true.
https://github.com/home-assistant/frontend/blob/6325e4493c1621e6e5865c2a1778f8956c6f1a08/src/panels/lovelace/hui-root.ts#L513

Should force be set to `this.lovelace.rebuild` here as well

**Comment by bramkragten** [comment]
@zsarnett Yeah, that sounds like the right solution. I tested it quickly and it did seem to work here, but I might have done it at the wrong place 🤔 

But if you change it to `force = this.lovelace!.rebuild;` it should be ok.

**Comment by timmo001** [comment]
Don't know if there have been changes since the gif in the description, but perhaps interactions with buttons etc. should be disabled in edit UI. You look to be able to trigger the UI quite easily there

**Comment by Misiu** [comment]
@zsarnett I see you are using muuri version 0.8.0.
I saw that a new version is almost ready (https://github.com/haltu/muuri/issues/373).
According to the info, it is stable and will be soon released. If fixes a lot of bugs and the performance is much better.
Maybe you could give version 0.9.0 a try?

**Comment by zsarnett** [comment]
Definitely will when it's released 😉

**Comment by bramkragten** [comment]
I will look at this end of this week 🙂

File: src/panels/lovelace/views/hui-view.ts (hunk 13 above)
Comment (bramkragten) [reply src/panels/lovelace/views/hui-view.ts:32]: import type

Comment (zsarnett) [reply src/panels/lovelace/views/hui-view.ts:32]: MAybe I can use the new import Type that was introduced?

**Comment by zsarnett** [comment]
Updating this PR for anyone who sees it. Bram and I have been talking back and forth and are planning to talk about the best way to handle edge cases and overall storing the configuration so that the views will always be in the order your stored when moving your cards. This is a large change that will take a very long time to make sure we get it right

**Comment by XtremeOwnageDotCom** [comment]
Any update? This would be a wonderful feature.

**Comment by zsarnett** [comment]
Still thinking of the best implementation. Exploring multiple options. I think this is the best option for ordering Cards just finding the correct library and way to do it

**Comment by XtremeOwnageDotCom** [comment]
I am not much of a front-end designer, and usually prefer to hang around the backend-
But, a few issues I can think of-

If you create a "state" for the elements in your view, while you are on your 1440p/4k screen, but, then open the same view on a 1080p screen, it is going to present issues. Likewise, for when you open it using a mobile view.

In my opinion, a good addition would be to maintain separate "states" for different viewport sizes. In the case of an application I work on at work, I don't have drag/drop functionality- however, I maintain a different "state" for each of the viewport sizes.

As an example, using bootstrap, https://getbootstrap.com/docs/4.4/layout/overview/

It supports 5 different "sizes". Perhaps taking the idea from them- store the view "state" depending on which of the "size groups" it fits into. So- if you are on a tablet with a small screen, it would display differently then your 4k screen on your computer.

Likewise- this is not the pull request for this- but, it would be wonderful if we were able to fully utilize large screen sizes, without having to build a "panel" dashboard composed of nesting vertical spans with horizational spans.

Just two cents, take it with a grain of salt. Front-end work isn't really my thing.

TLDR;

Save/Allow a different view "state" based on screen size.

**Comment by zsarnett** [comment]
The issue with doing that is its not very user friendly right. Its not intuitive and easy to create so many dashboards. It may be the way we go but my hope is to not have to or at least have the option to not do that and it still work

**Comment by zsarnett** [comment]
Something different coming soon 😃

**Comment by XtremeOwnageDotCom** [comment]
I eagerly await the upcoming results.

**Comment by fti7** [comment]
Any Updates on the "different" Solution? :-)

**Comment by lzinga** [comment]
Any updates on this?

**Comment by ScottG489** [comment]
@zsarnett Update? This is by far the most clunky thing in the UI at the moment.
